Replace all REST query empty result HTTP Status codes to 200

MarinX opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

All query params that are empty will now return success (http.StatusOK) instead of mixed http status codes (http.StatusNoContent/http.StatusBadRequest/http.StatusInternalServerError)

closes: #2007 

----

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [x] Added a relevant changelog entry: `sdkch add [section] [stanza] [message]`
- [x] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

fedekunze commented on x/gov/client/rest/rest.go:
```diff
@@ -310,7 +310,7 @@ func queryDepositHandlerFn(cdc *codec.Codec, cliCtx context.CLIContext) http.Han
 
 		if len(strProposalID) == 0 {
 			err := errors.New("proposalId required but not specified")
-			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
```
this is correct and shouldn't be replaced

MarinX replied:
noted. Reverted.

fedekunze commented on x/gov/client/rest/rest.go:
```diff
@@ -321,7 +321,7 @@ func queryDepositHandlerFn(cdc *codec.Codec, cliCtx context.CLIContext) http.Han
 
 		if len(bechDepositorAddr) == 0 {
 			err := errors.New("depositor address required but not specified")
-			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
```
ditto

MarinX replied:
noted. Reverted.

fedekunze commented on x/gov/client/rest/rest.go:
```diff
@@ -387,7 +387,7 @@ func queryVoteHandlerFn(cdc *codec.Codec, cliCtx context.CLIContext) http.Handle
 
 		if len(strProposalID) == 0 {
 			err := errors.New("proposalId required but not specified")
-			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
```
ditto

MarinX replied:
noted. Reverted.

fedekunze commented on x/gov/client/rest/rest.go:
```diff
@@ -398,7 +398,7 @@ func queryVoteHandlerFn(cdc *codec.Codec, cliCtx context.CLIContext) http.Handle
 
 		if len(bechVoterAddr) == 0 {
 			err := errors.New("voter address required but not specified")
-			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
```
ditto

MarinX replied:
noted. Reverted.

fedekunze commented on x/gov/client/rest/rest.go:
```diff
@@ -464,7 +464,7 @@ func queryVotesOnProposalHandlerFn(cdc *codec.Codec, cliCtx context.CLIContext)
 
 		if len(strProposalID) == 0 {
 			err := errors.New("proposalId required but not specified")
-			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
```
ditto

MarinX replied:
noted. Reverted.

fedekunze commented on x/gov/client/rest/rest.go:
```diff
@@ -579,7 +579,7 @@ func queryTallyOnProposalHandlerFn(cdc *codec.Codec, cliCtx context.CLIContext)
 
 		if len(strProposalID) == 0 {
 			err := errors.New("proposalId required but not specified")
-			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
```
ditto

MarinX replied:
noted. Reverted.

fedekunze commented on client/lcd/lcd_test.go:
```diff
@@ -114,7 +114,7 @@ func TestCoinSend(t *testing.T) {
 
 	// query empty
 	res, body := Request(t, port, "GET", fmt.Sprintf("/auth/accounts/%s", someFakeAddr), nil)
-	require.Equal(t, http.StatusNoContent, res.StatusCode, body)
```
we should add check for empty account (eg. `Account{}`)

MarinX replied:
I added a test `TestAccountBalanceQuery` which covers checking if we are returning empty Account{}

fedekunze commented on x/slashing/client/rest/query.go:
```diff
@@ -64,7 +59,7 @@ func signingInfoHandlerListFn(cliCtx context.CLIContext, storeName string, cdc *
 
 		_, page, limit, err := rest.ParseHTTPArgs(r)
 		if err != nil {
-			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
```
this is also correct and shouldn't be removed, since this checks for badly parsed query parameters

MarinX replied:
Noted. Reverted

fedekunze commented on x/slashing/client/rest/query.go:
```diff
@@ -81,7 +76,7 @@ func signingInfoHandlerListFn(cliCtx context.CLIContext, storeName string, cdc *
 
 		if len(validators.Validators) == 0 {
-			w.WriteHeader(http.StatusNoContent)
+			w.WriteHeader(http.StatusOK)
```
thanks ! we need to also test here that we're returning an empty array `[]` instead of `null`

MarinX replied:
I added a test `TestAccountBalanceQuery` which covers checking if we are returning []

fedekunze commented on x/slashing/client/rest/query.go:
```diff
@@ -100,7 +95,7 @@ func signingInfoHandlerListFn(cliCtx context.CLIContext, storeName string, cdc *
 
 		if len(signingInfoList) == 0 {
-			w.WriteHeader(http.StatusNoContent)
+			w.WriteHeader(http.StatusOK)
```
ditto

fedekunze commented on x/staking/client/rest/query.go:
```diff
@@ -154,7 +154,7 @@ func delegatorTxsHandlerFn(cliCtx context.CLIContext, cdc *codec.Codec) http.Han
 			actions = append(actions, staking.MsgBeginRedelegate{}.Type())
 			actions = append(actions, tags.ActionCompleteRedelegation)
 		default:
-			w.WriteHeader(http.StatusNoContent)
```
I guess we don't need to modify this since we'll probably delete this function since it's just a txs query (ie. duplicate)

MarinX replied:
Noted. Reverted

fedekunze commented on x/auth/client/rest/query.go:
```diff
@@ -49,7 +49,7 @@ func QueryAccountRequestHandlerFn(
 		// the query will return empty if there is no data for this account
 		if len(res) == 0 {
-			w.WriteHeader(http.StatusNoContent)
+			w.WriteHeader(http.StatusOK)
```
ditto, we need to test that this returns an empty `Account`

MarinX replied:
Added test which covers it `TestAccountBalanceQuery `

fedekunze commented on x/auth/client/rest/query.go:
```diff
@@ -88,7 +88,7 @@ func QueryBalancesRequestHandlerFn(
 		// the query will return empty if there is no data for this account
 		if len(res) == 0 {
-			w.WriteHeader(http.StatusNoContent)
+			w.WriteHeader(http.StatusOK)
```
curious if this function is a duplicate from the one above ?

MarinX replied:
Most of it is duplicated, only change is response. Eg. for `/auth/accounts/{address}` we are returning `Account{}` which in case `/bank/balances/{address}` we are returning `sdk.Coins`

The balance endpoint does not make sense for me, it should be `/auth/accounts/{address}/balances` and the same function should check the endpoint and return corresponding model. But this is a API breaking change so maybe it should be a separated issue for code cleanup :)

fedekunze replied:
I agree it doesn’t make much sense. Let’s open an issue for it

MarinX commented:
Summary of change:
- Reverted back most of statuses for empty query params
- For empty results, we are returning empty models
- Added `TestAccountBalanceQuery` which covers the case for empty results

MarinX commented:
noted.
Added the changelog entry

alessio commented on .pending/improvements/gaiarest/4065-http-200-with-e:
```diff
@@ -0,0 +1 @@
+4065-http-200-with-empty-results
```
This should start with `#2007` and contain and contain an explanation on _what_ has been fixed (not the _how_ or _why_)

MarinX replied:
Ahhh, noted! Thanks.
Updated

alexanderbez commented on .pending/improvements/gaiarest/4065-http-200-with-e:
```diff
@@ -0,0 +1 @@
+4065-http-200-with-empty-results
```
```suggestion
#4065 Return 200 status code on empty results
```

MarinX replied:
Thanks for the suggestion.
Updated

MarinX commented:
Thanks for the suggestion.
I updated it.
It should be ok now :)

alexanderbez commented:
@alessio please approve :)
